internal/miner/miner.go

```diff
@@ -62,34 +68,61 @@ func (m *Miner) Start() error {
 	return nil
 }
 
-func calculateHash(state State) (string, int) {
+func calculateHash(state State) (string, int64) {
 	nonce := state.Nonce
-	difficulty := state.DifficultyNumber
 
 	for {
-		// Construct the input using the state fields and nonce
-		input := fmt.Sprintf("%x%d%d%x%d%d",
-			nonce,
-			state.BlockNumber,
-			state.CurrentHash,
-			state.LeadingZeros,
-			state.DifficultyNumber,
-			state.EpochTime,
-		)
+		stateBytes, err := stateToBytes(state)
+		if err != nil {
+			logging.GetLogger().Error(err)
+			return "", nonce
+		}
 
 		// Calculate the hash
 		hasher := sha256.New()
-		hasher.Write([]byte(input))
+		hasher.Write(stateBytes)
 		hash := hasher.Sum(nil)
 
-		// Convert the hash to a hexadecimal string
-		hashStr := hex.EncodeToString(hash)
+		// Get the difficulty metrics for the hash
+		metrics := getDifficulty(hash)
 
-		// Check if the hash has the required number of leading zeros
-		if strings.HasPrefix(hashStr, strings.Repeat("00", difficulty)) {
-			return hashStr, nonce
+		// Check the condition
+		if metrics.LeadingZeros > state.LeadingZeros || (metrics.LeadingZeros == state.LeadingZeros && metrics.DifficultyNumber < 2) {
+			return hex.EncodeToString(hash), nonce
 		}
 
 		nonce++
+		state.Nonce = nonce
+	}
+}
+
+func stateToBytes(state State) ([]byte, error) {
+	buf := new(bytes.Buffer)
+	if err := binary.Write(buf, binary.LittleEndian, state); err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+func getDifficulty(hash []byte) DifficultyMetrics {
+	var metrics DifficultyMetrics
+	for indx, chr := range hash {
+		if chr != 0 {
+			if (chr & 0x0F) == chr {
```

Comment:
Magic numbers should have a comment and perhaps a `const`

Reply:
I told him this was OK in a chat. I agree that we should add a comment, at least. Still, let's not "block" this for it. :laughing: